File: Difficulty: Medium/Permutations of a String/permutations-of-a-string.java
class Solution {
    void solve(ArrayList<String>arr,StringBuilder s,int index)
    {
        if(index ==s.length())
        {
            arr.add(s.toString());
            return;
        }
        HashSet<Character>st = new HashSet<>();
        int n =s.length();
        for(int i= index;i<n;i++)
        {
        
            if(!st.contains(s.charAt(i)))
            {
                st.add(s.charAt(i));
                //swap 
                char temp = s.charAt(i);
                s.setCharAt(i,s.charAt(index));
                s.setCharAt(index,temp);
                //function call
                solve(arr,s,index+1);
                //ease the swap for the backtracking
                temp = s.charAt(i);
                s.setCharAt(i,s.charAt(index));
                s.setCharAt(index,temp);
            }
        }
    }
    public ArrayList<String> findPermutation(String s) {
        ArrayList<String>arr = new ArrayList<>();
        String res = "";
        StringBuilder s1 = new StringBuilder(s);
        solve(arr,s1,0);
        return arr;
    }
}

File: Difficulty: Medium/Permutations of a String/permutations-of-a-string.cpp
class Solution {
  public:
  void solve(string s,vector<string>&ans,int index)
  {
      if(index==s.length())
      {
          ans.push_back(s);
          return;
      }
      int n = s.length();
      unordered_set<char>st;
        for(int i = index;i<n;i++)
        {
            if(st.find(s[i])==st.end())
            {
                st.insert(s[i]);
                swap(s[i],s[index]);
                solve(s,ans,index+1);
                swap(s[i],s[index]);
            }
        }
  }
    vector<string> findPermutation(string &s) {
        vector<string>ans;
        solve(s,ans,0);
        return ans;
    }
};
